api/src/services/uploads/uploads.ts

```diff
@@ -147,3 +147,114 @@ export const Upload: UploadRelationResolvers = {
     return latestValidation
   },
 }
+
+/*
+This function should return the most recent upload for each expenditure category and agency for the `currentUser`'s organization.
+The uploads must be grouped by expenditure category and agency. If there are multiple uploads for the grouping, the most recent upload must be chosen.
+The S3 Object key must be set on each upload object.
+The return value structure should look like the following:
+{
+  EC-Code: {
+    AgencyId: {
+      UploadObject
+    }
+  }
+}
+Example:
+{
+  '1A': {
+    '1': {
+      id: 1,
+      filename: 'file1.csv',
+      objectKey: 'uploads/1/1/1/1/file1.csv',
+      ...
+    },
+    '2': {
+      id: 2,
+      filename: 'file2.csv',
+      objectKey: 'uploads/1/2/1/2/file2.csv',
+      ...
+    },
+    ...
+  },
+  '1B': {
+    '1': {
+      id: 3,
+      filename: 'file3.csv',
+      objectKey: 'uploads/1/1/1/3/file3.csv',
+      ...
+    },
+    ...
+  },
+  ...
+}
+*/
+export const getUploadsByExpenditureCategory = async () => {
+  const organization = await db.organization.findFirst({
+    where: { id: context.currentUser.agency.organizationId },
+  })
+
+  const reportingPeriod = await db.reportingPeriod.findFirst({
+    where: { id: organization.preferences['current_reporting_period_id'] },
+  })
+
+  // valid upload validations
+  const validUploadIds = await db.uploadValidation.findMany({
+    where: {
+      passed: true,
+      upload: { agency: { organizationId: organization.id } },
+    },
+    select: {
+      uploadId: true,
+    },
+    distinct: ['uploadId'],
+  })
+
+  const uploadsForPeriod = await db.upload.findMany({
+    where: {
+      reportingPeriodId: reportingPeriod.id,
+      agency: { organizationId: organization.id },
+      id: { in: validUploadIds.map((v) => v.uploadId) },
+    },
+    include: { expenditureCategory: true },
+  })
+
+  const uploadsByExpenditureCategory = {}
+
+  // Get the most recent upload for each expenditure category and agency and set the S3 Object key
+  for (const upload of uploadsForPeriod)  {
```

Comment:
Ultimately up to you and totally fine in a follow-up, but wondering if a small wrapper class for `uploadsByExpenditureCategory` with some getters and setters might make it easier for future devs to understand / edit this code?

Reply:
Good idea, but let's do this in a follow up PR. I think this PR is a blocker on some other work.